### ISSUE TITLE
Incorrect log_probs dim causes a confusing error about NaN losses on CTC loss/search

### ISSUE THREAD

**Issue by AsuMagic** [speechbrain/speechbrain#1589]
## Summary

Accidentally undersizing the output vector for a CTC-based model (and possibly others) may produce a confusing NaN loss error after a couple of batches, which is not helpful in finding the root cause.  
PyTorch's `detect_anomaly` only points as NaNs being found in `log_softmax`, which does not help either.

Ideally, SB could instead detect this issue and produce a more helpful error/warning/hint when this occurs.

## Context

(This is mostly skippable, but should help making the issue more googleable and to understand the developer experience issue)

While building an ASR recipe based on the LibriSpeech wav2vec2-based CTC recipe, I have encountered a surprising issue where my training would crash due to a NaN loss just after a couple of batches.  
After some shotgun debugging, I found out that I had completely overlooked **the `output_neurons` hyperparameter, which was set way too low for my dataset**, which contains many more output tokens than LibriSpeech. This apparently didn't result in any meaningful warning or error.  
Incidentally, it seems that a colleague has run into the exact same issue, but that has not been confirmed yet.

## Possible fix

My suggestion is that all relevant loss functions in `nnet.losses` (CTC loss in my case) could include sanity checks to ensure that the prediction tensor is large enough to accept all possible targets. The same could be applied to `ctc_greedy_decode` (which is only used at inference time), and possibly to other functions in SB that could exhibit the same issue that are to be investigated.

That being said, I have some concerns around this but I don't have enough knowledge right now to know whether they're an issue in practice, or to know what the best course of action is:
- The parameters to `ctc_loss` do not seem to help knowing how many tokens should exist. It could be possible to check for the `max` of `targets` but that's a little silly and means we don't catch the issue as early as possible.
- I am not sure whether adding such a check could cause unintended side effects. I cannot think of one right now, though.

**Comment by TParcollet** [comment]
The real issue is that in SB, we do not require the user to identify a layer as an output layer. A quick solution, imho, would be to add a really visible log output explicitons the size of the vocabulary. This error is very very very common among our community. 

An other possibility would be to add a check in our SentencePiece interface. We could ask for a mandatory argument "vocab_size" and check if what we produce actually is of this size. Imho this would be even better.

**Comment by Adel-Moumen** [comment]
@AsuMagic are you willing to do a PR with the solution mentioned by @TParcolled, please? That would be awesome! Thanks  :-)

**Comment by AsuMagic** [comment]
> An other possibility would be to add a check in our SentencePiece interface. We could ask for a mandatory argument "vocab_size" and check if what we produce actually is of this size. Imho this would be even better.

`speechbrain.tokenizers.SentencePiece.SentencePiece` itself already requires a `vocab_size`, as it is required for tokenizer training.  
Passing an incorrect `vocab_size` to it could cause problems down the line too, but I don't think there is anything that can be done at the `SentencePiece` class level.

This is not what is used in the CTC case, however. `CTCTextEncoder` is used instead, meaning the issue originates from `CategoricalEncoder`/`TextEncoder`. There is no obvious single place to insert such a check, however. You aren't forced to "finalize" an encoder, and you're not always done modifying it at `load()` time (even if that may not be a great idea). In other words, it seems you can only do such a check by the point you encode a label.

For the sake of backwards compatibility, and as to avoid making the check too annoying, my idea is:

- Add a `ensure_vocab_size(vocab_size)` method. Make it immediately check whether the vocabulary size is correct. If it is not, then a hard error is thrown.
- Add a `ignore_vocab_size()` method.
- When encoding a label _for the first time_ (all methods seem to go through `encode_label`), and _neither_ of `ensure_vocab_size` or `ignore_vocab_size` have been called, then log a `UserWarning` that goes along the lines of:

```
You have never sanity checked for the vocabulary size for [encoder class name].
Please use `tokenizer.ensure_vocab_size` (e.g. usually `tokenizer.ensure_vocab_size(hparams["output_neurons"])` to make sure the token list exactly matches the length you expect from your dataset and code.
Some code (such as some loss functions) will never warn you, which often lead to models failing to train properly.
If you are sure this does not apply to you, you may use `tokenizer.ignore_vocab_size()`.
```

I'm not sure if multiline or such verbose warnings are good practice but the details seem important to me.

Generally speaking, aside of additional sanity checking in loss functions, I don't think there is a completely fool-proof way of checking this. Tokenizers have no visibility of what you do with its output nor can it know whether the `vocab_size` you passed is good or bad.  
However, the mere fact of enforcing or warning about checking the vocabulary size should catch most user errors.

**Comment by Adel-Moumen** [comment]
@AsuMagic I think it does make sense.

**Comment by Adel-Moumen** [comment]
Fixed in #1852
